File: leetcode_problems/python/lc658_find_k_closest_elements.py
class Solution:
    def show(
        self,
        low: int,
        high: int,
        mid: int,
    ) -> None:
        print("low:", low, "high:", high, "mid:", mid)

    def findClosestElements(
        self,
        arr: list[int],
        k: int,
        x: int,
    ) -> list[int]:
        def good(i: int):
            return x - arr[i] > arr[i + k] - x

        low = -1
        high = len(arr) - k
        while high - low > 1:
            mid = low + (high - low) // 2
            self.show(low, high, mid)
            if good(mid):
                low = mid
                print("low changed:", low)
            else:
                high = mid
                print("high changed:", high)

        return arr[high : high + k]


if __name__ == "__main__":
    arr = [1, 1, 2, 2, 3, 4, 5, 5, 6, 6, 7, 8, 9, 9, 10]
    k = 5
    x = 10

    print(len(arr))
    print(Solution().findClosestElements(arr, k, x))

File: leetcode_problems/go/lc658_find_k_closest_elements.go
package main

func findClosestElementsBS(arr []int, x int) (int, int) {
	l := 0
	h := len(arr)
	for h-l > 1 {
		m := l + (h-l)/2
		if arr[m] < x {
			l = m
		} else {
			h = m
		}
	}
	return l, h
}

func findClosestElements(arr []int, k int, x int) []int {
	n := len(arr)
	p1, p2 := findClosestElementsBS(arr, x)
	for k > 0 {
		if p1 < 0 {
			p2++
		} else if p2 >= n {
			p1--
		} else if abs(arr[p1]-x) == abs(arr[p2]-x) && arr[p1] < arr[p2] {
			p1--
		} else if abs(arr[p1]-x) == abs(arr[p2]-x) && arr[p1] > arr[p2] {
			p2++
		} else if abs(arr[p1]-x) < abs(arr[p2]-x) {
			p1--
		} else {
			p2++
		}
		k--
	}
	return arr[p1+1 : p2]
}
